src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,651 @@
+                              ParseInput& in,
+                              Name name,
+                              std::optional<ImportNames> importNames);
+template<typename Ctx>
+Result<> parseDefs(Ctx& ctx,
+                   std::string_view input,
+                   const std::vector<DefPos>& defs,
+                   MaybeResult<> (*parser)(Ctx&, ParseInput&));
+
+// =====
+// Types
+// =====
+
+// numtype ::= 'i32' => i32
+//           | 'i64' => i64
+//           | 'f32' => f32
+//           | 'f64' => f64
+// vectype ::= 'v128' => v128
+// valtype ::= t:numtype => t
+//           | t:vectype => t
+//           | t:reftype => t
+template<typename Ctx>
+Result<typename Ctx::ValTypeT> valtype(Ctx& ctx, ParseInput& in) {
+  if (in.takeKeyword("i32"sv)) {
+    RETURN_OR_OK(Type::i32);
+  } else if (in.takeKeyword("i64"sv)) {
+    RETURN_OR_OK(Type::i64);
+  } else if (in.takeKeyword("f32"sv)) {
+    RETURN_OR_OK(Type::f32);
+  } else if (in.takeKeyword("f64"sv)) {
+    RETURN_OR_OK(Type::f64);
+  } else if (in.takeKeyword("v128"sv)) {
+    RETURN_OR_OK(Type::v128);
+  } else {
+    // TODO: reftype
+    return in.err("expected valtype");
+  }
+}
+
+// globaltype ::= t:valtype               => const t
+//              | '(' 'mut' t:valtype ')' => var t
+template<typename Ctx>
+Result<typename Ctx::GlobalTypeT> globaltype(Ctx& ctx, ParseInput& in) {
+  // '(' 'mut' t:valtype ')'
+  if (in.takeSExprStart("mut"sv)) {
+    auto type = valtype(ctx, in);
+    CHECK_ERR(type);
+
+    if (!in.takeRParen()) {
+      return in.err("expected end of globaltype");
+    }
+    RETURN_OR_OK((GlobalType{Mutable, *type}));
+  }
+
+  // t:valtype
+  auto type = valtype(ctx, in);
+  CHECK_ERR(type);
+  RETURN_OR_OK((GlobalType{Immutable, *type}));
+}
+
+// =======
+// Modules
+// =======
+
+// ('(' 'import' mod:name nm:name ')')?
+MaybeResult<ImportNames> inlineImport(ParseInput& in) {
+  if (!in.takeSExprStart("import"sv)) {
+    return {};
+  }
+  auto mod = in.takeName();
+  if (!mod) {
+    return in.err("expected import module");
+  }
+  auto nm = in.takeName();
+  if (!nm) {
+    return in.err("expected import name");
+  }
+  if (!in.takeRParen()) {
+    return in.err("expected end of import");
+  }
+  return {{*mod, *nm}};
+}
+
+// ('(' 'export' name ')')*
+Result<std::vector<Name>> inlineExports(ParseInput& in) {
+  std::vector<Name> exports;
+  while (in.takeSExprStart("export"sv)) {
+    auto name = in.takeName();
+    if (!name) {
+      return in.err("expected export name");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of import");
+    }
+    exports.push_back(*name);
+  }
+  return exports;
+}
+
+// global ::= '(' 'global' id? ('(' 'export' name ')')* gt:globaltype e:expr ')'
+//          | '(' 'global' id? '(' 'import' mod:name nm:name ')'
+//                gt:globaltype ')'
+template<typename Ctx> MaybeResult<> global(Ctx& ctx, ParseInput& in) {
+  [[maybe_unused]] auto pos = in.getPos();
```

Comment:
This is actually used, so why is this annotation needed?

We also have `WASM_UNUSED` in the codebase currently.

Reply:
It's only used in one arm of the `if constexpr`, so I think one of the compilers on CI gets confused and emits an error when specializing the template such that it doesn't include that arm. Since it's not actually unused, I would rather not use `WASM_UNUSED`.